Convert template objects to singletons

kpdecker opened the pull request:
Create file-scoped template objects that are instantiated once and reused between different tagged template evaluations. This is a closer match to the spec behavior which demands reuse, but does not match the spec exactly with respect to scoping as it’s possible to have similar template objects defined in different file scopes within the same program. For now we are not able to cleanly handle this case as it would require a registry or similar lookup mechanism that is consistent across different combinations of builds and optimizers.

This has a 40x increase in six-speed throughput.

Partial fix for #971

sebmck commented on test/core/fixtures/esnext/es6-templates/tagged-template-expression.js:
```diff
@@ -24,3 +25,7 @@ function tagInterpolateLast(strings) {
+
+// ENsure that we don't have conflicts between distinct templates with similar strings.
+assert.equal(String.raw `${ 42 }b`, '42b');
+assert.equal(String.raw `b${ 42 }`, 'b42');
```
This is the wrong place for these tests. esnext contains regressions tests from the [esnext merger](http://babeljs.io/blog/2015/01/12/6to5-esnext/).

kpdecker replied:
Where would a better location be?

kpdecker commented:
@sebmck reworked/fixed tests and rebased. Also added explicit block hoist value to prevent potential future error if a helper uses IIFE or other non-function expression helper as these wouldn't be hoisted by the runtime. Using arbitrary `1.9` constant feels a bit hacky, but I wasn't sure of any other good way of ensuring the helper dependency was properly init.

sebmck commented:
Sorry @kpdecker, just an update but I haven't forgotten about this. Likely wont get around to having a decent look at this until Monday. Thank you so much for the contribution though!

sebmck commented on src/babel/transformation/file/index.js:
```diff
@@ -413,6 +413,30 @@ export default class File {
+    var stringIds = raw.elements.map(function(string) {
+      return string.value;
+    });
+    var name = `${helperName}_${raw.elements.length}_${stringIds.join(",")}`;
```
This will cause extremely long variable names for big strings.

kpdecker replied:
Yes. My assumption was this was a non-issue as the minimizer will optimize.
This of course assumes that you're using a minimizer... Perhaps we should
save a look up table on the file instance to shorter names?

On Tue, Jul 7, 2015 at 11:56 AM Sebastian McKenzie <notifications@github.com>
wrote:

> In src/babel/transformation/file/index.js
> <https://github.com/babel/babel/pull/1878#discussion_r34024402>:
>
> > @@ -413,6 +413,30 @@ export default class File {
> >      return uid;
> >    }
> >
> > +  addTemplateObject(helperName: string, strings: Array, raw: Array): Object {
> > +    // Generate a unique name based on the string literals so we dedupe
> > +    // identical strings used in the program.
> > +    var stringIds = raw.elements.map(function(string) {
> > +      return string.value;
> > +    });
> > +    var name = `${helperName}_${raw.elements.length}_${stringIds.join(",")}`;
>
> This will cause extremely long variable names for big strings.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/pull/1878/files#r34024402>.
>

sebmck replied:
Yeah, a lookup table is probably the best bet. All that would be required actually would be a different argument given to `generateUidIdentifier` since `this.declarations` is already being used with the complete key.

kpdecker replied:
Will look into it. (Currently on vacation so might take a little bit)
On Tue, Jul 7, 2015 at 12:55 PM Sebastian McKenzie <notifications@github.com>
wrote:

> In src/babel/transformation/file/index.js
> <https://github.com/babel/babel/pull/1878#discussion_r34028273>:
>
> > @@ -413,6 +413,30 @@ export default class File {
> >      return uid;
> >    }
> >
> > +  addTemplateObject(helperName: string, strings: Array, raw: Array): Object {
> > +    // Generate a unique name based on the string literals so we dedupe
> > +    // identical strings used in the program.
> > +    var stringIds = raw.elements.map(function(string) {
> > +      return string.value;
> > +    });
> > +    var name = `${helperName}_${raw.elements.length}_${stringIds.join(",")}`;
>
> Yeah, a lookup table is probably the best bet. All that would be required
> actually would be a different argument given to generateUidIdentifier
> since this.declarations is already being used with the complete key.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/pull/1878/files#r34028273>.
>

sebmck replied:
All good, no rush at all. Sorry for taking so long to get to this.

On Tuesday, 7 July 2015, Kevin Decker <notifications@github.com> wrote:

> In src/babel/transformation/file/index.js
> <https://github.com/babel/babel/pull/1878#discussion_r34029342>:
>
> > @@ -413,6 +413,30 @@ export default class File {
> >      return uid;
> >    }
> >
> > +  addTemplateObject(helperName: string, strings: Array, raw: Array): Object {
> > +    // Generate a unique name based on the string literals so we dedupe
> > +    // identical strings used in the program.
> > +    var stringIds = raw.elements.map(function(string) {
> > +      return string.value;
> > +    });
> > +    var name = `${helperName}_${raw.elements.length}_${stringIds.join(",")}`;
>
> Will look into it. (Currently on vacation so might take a little bit)
> … <#14e6838d0d0339d7_>
> On Tue, Jul 7, 2015 at 12:55 PM Sebastian McKenzie <
> notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote: In
> src/babel/transformation/file/index.js <
> https://github.com/babel/babel/pull/1878#discussion_r34028273>: > @@
> -413,6 +413,30 @@ export default class File { > return uid; > } > > +
> addTemplateObject(helperName: string, strings: Array, raw: Array): Object {
> > + // Generate a unique name based on the string literals so we dedupe > +
> // identical strings used in the program. > + var stringIds =
> raw.elements.map(function(string) { > + return string.value; > + }); > +
> var name = `${helperName}_${raw.elements.length}_${stringIds.join(",")}`;
> Yeah, a lookup table is probably the best bet. All that would be required
> actually would be a different argument given to generateUidIdentifier since
> this.declarations is already being used with the complete key. — Reply to
> this email directly or view it on GitHub <
> https://github.com/babel/babel/pull/1878/files#r34028273>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/pull/1878/files#r34029342>.
>


-- 
Sebastian McKenzie

kpdecker commented:
Updated to use `_templateObject` and derived names and also updated for the latest code in master.

sebmck commented on packages/babel/src/traversal/scope/index.js:
```diff
@@ -829,14 +829,15 @@ export default class Scope {
 
     var unique = opts.unique;
     var kind   = opts.kind || "var";
+    var blockHoist = opts._blockHoist || 2;
```
This will need to be `opts._blockHoist == null ? 2 : opts._blockHoist` in case it's `0`.

sebmck commented:
Sorry for the delay @kpdecker, could've sworn I'd already merged this. The only thing is the `_blockHoist` check and this is GTM. Thank you! :blush:

kpdecker commented:
Updated

sebmck commented:
Thanks!
